tests/unit_tests/carrot_convergence.cpp

```diff
@@ -37,470 +37,351 @@
 #include "string_tools.h"
 
 using namespace carrot;
-using epee::string_tools::hex_to_pod;
-using epee::string_tools::pod_to_hex;
+
+namespace
+{
+bool operator==(const mx25519_pubkey &lhs, const mx25519_pubkey &rhs)
+{
+    return 0 == memcmp(lhs.data, rhs.data, sizeof(mx25519_pubkey));
+}
+
+template <typename T>
+const T &pod_unwrap(const T &v) { return v; }
+template <typename T>
+T &pod_unwrap(T &v) { return v; }
+const crypto::ec_scalar &pod_unwrap(const crypto::secret_key &v) { return v; }
+crypto::ec_scalar &pod_unwrap(crypto::secret_key &v) { return v; }
+
+template <typename T>
+std::string pod_to_hex(const T &v)
+{
+    return epee::string_tools::pod_to_hex(pod_unwrap(v));
+}
+
+template <typename T>
+T hex_to_pod(const boost::string_ref hex)
+{
+    T val{};
+    if (!epee::string_tools::hex_to_pod(hex, pod_unwrap(val)))
+        throw std::runtime_error("hex failed to decode to value");
+    return val;
+}
+
+template <typename T>
+testing::AssertionResult matches_hex(const boost::string_ref expected_hex, const T &actual) {
+    if (expected_hex.size() != sizeof(T) * 2)
+        throw std::logic_error("Expected hex is wrong size for type");
+
+    const T expected = hex_to_pod<T>(expected_hex);
+    const std::string actual_hex = pod_to_hex(actual);
+
+    const bool matches = expected == actual;
+    const bool matches_hex = expected_hex == actual_hex;
+    if (matches != matches_hex)
+        throw std::logic_error("Type is not practically POD-like");
+
+    if (!matches)
+    {
+        return testing::AssertionFailure() << "actual hex is " << actual_hex;
+    }
+
+    return testing::AssertionSuccess();
+}
+
+template <typename T>
+struct hex_value_t
+{
+    hex_value_t(const boost::string_ref hex):
+        hex(hex),
+        value(hex_to_pod<T>(hex))
+    {}
+
+    const std::string hex;
+    const T value;
+};
+
+static const hex_value_t<crypto::secret_key> s_master("6e02e67b303dc713276bb1a4d70b0083b78e4f50e34e209da9f0377cdc3d376e"); // chosen by fair dice roll.
+                                                                                                                     // guaranteed to be random.
+static const hex_value_t<crypto::secret_key> k_prove_spend("c9651fc906015afeefdb8d3bf7be621c36e035de2a85cb22dd4b869a22086f0e");
+static const hex_value_t<crypto::public_key> partial_spend_pubkey("eef3184e91505660c8ccbdeec1bd3b1b7b56d2c39efcad8a036f963470d6f498");
+static const hex_value_t<crypto::secret_key> s_view_balance("59b2ee8646923309384704613418f5982b0167eb3cd87c6c067ee10700c3af91");
+static const hex_value_t<crypto::secret_key> s_generate_image_preimage("e36e8a1bc422056597ce3f629611e410185e580557f81726b8979006cee3420b");
```

Comment:
`s_generate_image_preimage` and `k_generate_image` are the same, they must not be :)

I assume it's the same issue found on https://github.com/seraphis-migration/monero/pull/245#pullrequestreview-3531489832

On my own implementation, got `0f3bf96a0642ab4cd10e8c64fba1cc535379ec18dbc7d304d50eb753197e266f` and `dabc1ed54dc44f68f67200a1a66ee30b3237f05c2f6dc0dd47e5743431ac800b` respectively.